cli/main.c

```diff
@@ -1071,6 +1071,134 @@ static int fw_update(int argc, char **argv, struct command *cmd,
 	return ret;
 }
 
+static int mrpc_ses( int argc, char **argv, struct command *cmd, struct plugin *plugin )
+{
+    int         ret  = 0;
+    const char *desc = "Send an MRPC SES command to the firmware";
+
+    static struct
+    {
+        struct switchtec_dev *dev;
+        FILE                 *fp;
+    }   cfg = { 0 };
+
+    const struct argconfig_options opts[] =
+    {
+        DEVICE_OPTION,
+        {
+            "mrpc_ses_input_file",
+            .cfg_type      = CFG_FILE_R,
+            .value_addr    = &cfg.fp,
+            .argument_type = required_positional,
+            .help          = "MRPC SES file to use as input"
+        },
+        { NULL }
+    };
+
+    argconfig_parse( argc, argv, desc, opts, &cfg, sizeof(cfg) );
+
+    ret = switchtec_send_mrpc_ses_command( cfg.dev, cfg.fp );
+
+    fclose( cfg.fp );
+
+    if( ret )
+    {
+        printf( "Error sending MRPC SES command.\n" );
+    }
+
+    return ret;
+}
+
+uint8_t ascii_to_bin( char input )
+{
+    if( (input >= 0x30) && (input <= 0x39) )
+    {
+        return ( (uint8_t)input - 0x30 );
+    }
+
+    if( (input >= 0x41) && (input <= 0x46) )
+    {
+        return ( (uint8_t)input - 0x37 );
+    }
+
+    if( (input >= 0x61) && (input <= 0x66) )
+    {
+        return ( (uint8_t)input - 0x57 );
+    }
+
+    return ( 0xFF );
+}
+
+int switchtec_send_mrpc_ses_command(struct switchtec_dev *dev, FILE *fp)
+{
+    int      ret = 0;
+    uint8_t  mrpc_input[MRPC_MAX_DATA_LEN];
+    uint8_t  mrpc_output[MRPC_MAX_DATA_LEN];
+    uint8_t  byte;
+    uint8_t  byte_0;
+    uint8_t  byte_1;
+    uint8_t  byte_2;
+    uint8_t  byte_3;
+    uint32_t dword;
+    uint32_t index = 0;
+    uint32_t exit  = 0;
+    char     input;
+
+    memset(mrpc_input,  0, sizeof(mrpc_input));
+    memset(mrpc_output, 0, sizeof(mrpc_output));
+
+    do
+    {
+        input = (char)fgetc( fp );
+
+        if( input == EOF )
+        {
+            exit = 1;
+        }
+        else
+        {
+            byte = ascii_to_bin( input );
+
+            if( byte != 0xFF )
+            {
+                mrpc_input[index] = byte << 4;
+
+                input = (char)fgetc( fp );
+                byte  = ascii_to_bin( input );
+                mrpc_input[index] |= byte;
+                index++;
+            }
+        }
+
+    }   while( (exit == 0) && (index < MRPC_MAX_DATA_LEN) );
+
```

Comment:
This is an unacceptable interface. Expecting the user to construct hand crafted hex files and then have to parse the result is way too complicated. It's too reminiscent of the psx_app era. Everything in switchtec-user must largely be usable without needing to comb through the switch's device specification document.